### ISSUE TITLE
[LDAP] Failure to account for non-posix group(s)

### ISSUE THREAD

**Issue by FroggyFlox** [rockstor/rockstor-core#2673]
Thanks to forum member **bhotrock** for not only bringing this issue to light, but for their time and effort in debugging and verifying cause of issue; see thread below:
[https://forum.rockstor.com/t/ldap-service-breaks-users-and-shares-interface/9014](https://forum.rockstor.com/t/ldap-service-breaks-users-and-shares-interface/9014?u=flox)

When visiting any page that fetches currents users, the following error occurs when Rockstor has joined an LDAP server with non-posix groups.

```
Internal Server Error: /api/users
Traceback (most recent call last):
  File "/opt/rockstor/src/rockstor/storageadmin/models/user.py", line 59, in groupname
    groupname = grp.getgrgid(self.gid).gr_name
KeyError: 'getgrgid(): gid not found: 10000'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
(...)
  File "/opt/rockstor/src/rockstor/storageadmin/models/user.py", line 64, in groupname
    "ifp_groups", int(self.gid), "name"
  File "/opt/rockstor/src/rockstor/system/users.py", line 311, in ifp_get_properties_from_name_or_id
    my_obj = bus.get_object(ifp_bus_name, ifp_iface.FindByID(target))
(...)
  File "/opt/rockstor/.venv/lib64/python3.6/site-packages/dbus/connection.py", line 653, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: sbus.Error.NotFound: No such file or directory
[13/Sep/2023 16:35:31] "GET /api/users?page=1&format=json&page_size=32000&count= HTTP/1.1" 500 38609

```

<details><summary>Full error</summary>

```
Internal Server Error: /api/users
Traceback (most recent call last):
  File "/opt/rockstor/src/rockstor/storageadmin/models/user.py", line 59, in groupname
    groupname = grp.getgrgid(self.gid).gr_name
KeyError: 'getgrgid(): gid not found: 10000'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/rest_framework/views.py", line 495, in dispatch
    response = self.handle_exception(exc)
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/rest_framework/views.py", line 455, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/rest_framework/views.py", line 466, in raise_uncaught_exception
    raise exc
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/rest_framework/views.py", line 492, in dispatch
    response = handler(request, *args, **kwargs)
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/rest_framework/generics.py", line 241, in get
    return self.list(request, *args, **kwargs)
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/rest_framework/mixins.py", line 45, in list
    return self.get_paginated_response(serializer.data)
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/rest_framework/serializers.py", line 768, in data
    ret = super(ListSerializer, self).data
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/rest_framework/serializers.py", line 262, in data
    self._data = self.to_representation(self.instance)
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/rest_framework/serializers.py", line 686, in to_representation
    self.child.to_representation(item) for item in iterable
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/rest_framework/serializers.py", line 686, in <listcomp>
    self.child.to_representation(item) for item in iterable
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/rest_framework/serializers.py", line 517, in to_representation
    attribute = field.get_attribute(instance)
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/rest_framework/fields.py", line 453, in get_attribute
    return get_attribute(instance, self.source_attrs)
  File "/opt/rockstor/.venv/lib/python3.6/site-packages/rest_framework/fields.py", line 101, in get_attribute
    instance = getattr(instance, attr)
  File "/opt/rockstor/src/rockstor/storageadmin/models/user.py", line 64, in groupname
    "ifp_groups", int(self.gid), "name"
  File "/opt/rockstor/src/rockstor/system/users.py", line 311, in ifp_get_properties_from_name_or_id
    my_obj = bus.get_object(ifp_bus_name, ifp_iface.FindByID(target))
  File "/opt/rockstor/.venv/lib64/python3.6/site-packages/dbus/proxies.py", line 72, in __call__
    return self._proxy_method(*args, **keywords)
  File "/opt/rockstor/.venv/lib64/python3.6/site-packages/dbus/proxies.py", line 147, in __call__
    **keywords)
  File "/opt/rockstor/.venv/lib64/python3.6/site-packages/dbus/connection.py", line 653, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: sbus.Error.NotFound: No such file or directory
[13/Sep/2023 16:35:31] "GET /api/users?page=1&format=json&page_size=32000&count= HTTP/1.1" 500 38609
```

</details> 

This was with the following LDAP user setup (no group exists on the LDAP server yet):
```
# Create ou=users
dn: ou=users,@LDAP_BASE_DN@
objectClass: organizationalUnit
ou: users

# Create rockuser01
dn: uid=rockuser01,ou=users,@LDAP_BASE_DN@
cn: rockuser01
gidnumber: 100
homedirectory: /home/rockuser01
objectclass: account
objectclass: posixAccount
objectclass: top
uid: rockuser01
uidnumber: 20000
userpassword: user

# Create rockuser02
dn: uid=rockuser02,ou=users,@LDAP_BASE_DN@
cn: rockuser02
gidnumber: 10000
homedirectory: /home/rockuser02
objectclass: account
objectclass: posixAccount
objectclass: top
uid: rockuser02
uidnumber: 10000
userpassword: user10
```

We can thus see that:
- all went well for `rockuser01` as it has a `gidNumber` of 100, which exists on the Rockstor machine.
- failure with `rockuser02`: `groupname = grp.getgrgid(self.gid).gr_name` failed with a `KeyError` as expected as it is a LDAP user, but then infopipe failed to also retrieve this record:
```
  File "/opt/rockstor/src/rockstor/system/users.py", line 311, in ifp_get_properties_from_name_or_id
    my_obj = bus.get_object(ifp_bus_name, ifp_iface.FindByID(target))
(...)
dbus.exceptions.DBusException: sbus.Error.NotFound: No such file or directory
```
This failure can also be observed when manually looking up for that gid:
```
buildvm155:/opt/rockstor # poetry run dbus-send --system --print-reply  --dest=org.freedesktop.sssd.infopipe /org/freedesktop/sssd/infopipe/Groups org.freedesktop.sssd.infopipe.Groups.FindByID uint32:"100"
Error sbus.Error.NotFound: No such file or directory

buildvm155:/opt/rockstor # poetry run dbus-send --system --print-reply  --dest=org.freedesktop.sssd.infopipe /org/freedesktop/sssd/infopipe/Groups org.freedesktop.sssd.infopipe.Groups.FindByID uint32:"10000"
Error sbus.Error.NotFound: No such file or directory
```

This is actually **normal** behavior as no posix account exists for that `gidNumber` on the LDAP server. We should thus account for this scenario.

**Comment by FroggyFlox** [comment]
As a confirmation, set the LDAP users either to a local (rockstor) group, or a LDAP posixAccount group as follows
```
# Create ou=users
dn: ou=users,@LDAP_BASE_DN@
objectClass: organizationalUnit
ou: users

# Create rockuser01
dn: uid=rockuser01,ou=users,@LDAP_BASE_DN@
cn: rockuser01
gidnumber: 100
homedirectory: /home/rockuser01
objectclass: account
objectclass: posixAccount
objectclass: top
uid: rockuser01
uidnumber: 20000
userpassword: user

# Create rockuser02
dn: uid=rockuser02,ou=users,@LDAP_BASE_DN@
cn: rockuser02
gidnumber: 21000
homedirectory: /home/rockuser02
objectclass: account
objectclass: posixAccount
objectclass: top
uid: rockuser02
uidnumber: 10000
userpassword: user10


# Create ou=groups
dn: ou=groups,@LDAP_BASE_DN@
objectClass: organizationalUnit
objectClass: top
ou: groups
description: Organizational unit for user groups

# Creating ldapgroup1 group in the groups unit
dn: cn=ldapgroup1,ou=groups,@LDAP_BASE_DN@
cn: ldapgroup1
gidNumber: 21000
objectClass: top
objectClass: posixGroup
description: LDAP Group
```
```
# Add a new member to ldagroup1 group
dn: cn=ldapgroup1,ou=groups,@LDAP_BASE_DN@
changetype: modify
add: memberUid
memberUid: rockuser02
```
Then, navigate to the Users page: no error:
![image](https://github.com/rockstor/rockstor-core/assets/30297881/813c37d2-57cc-46e8-87ca-3a73cafe20fc)

**Comment by FroggyFlox** [comment]
We already return `None` during our user lookup when:
- `system.users.ifp_get_properties_from_name_or_id()` is called but SSSD is not running:
https://github.com/rockstor/rockstor-core/blob/b4128ad92f7b86597490b890d08ae9dc8139de3f/src/rockstor/system/users.py#L300-L301

- the `groupname()` property of the `User` model cannot get the information:
https://github.com/rockstor/rockstor-core/blob/b4128ad92f7b86597490b890d08ae9dc8139de3f/src/rockstor/storageadmin/models/user.py#L46-L68

It does seem safe to also return `None` when `system.users.ifp_get_properties_from_name_or_id()` when the group record cannot be found.

**Comment by phillxnet** [comment]
Closing as:
Fixed by #2674
